resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/dtree/AbstractDataTree.java

```diff
@@ -213,10 +213,20 @@ public boolean isImmutable() {
 	/**
 	 * Returns the key of the root node.
 	 */
-	public IPath rootKey() {
+	public static final IPath rootKey() {
 		return Path.ROOT;
 	}
 
+	public boolean isRoot(IPath key) {
+		return key == rootKey();
+	}
+
+	private static final IPath[] ROOT_PATHS = new IPath[] { rootKey() };
+
+	public IPath[] rootPaths() {
+		return ROOT_PATHS;
```

Comment:
Since array content is modifiable, any caller could per mistake change the array.
Not sure if this is a good idea to provide this method in the parent class.
I see it is only used in `org.eclipse.core.internal.watson.ElementTree.getChildIDs(IPath)`.
Why not to add that there, as private method and init it to the instance (not static) array?

Reply:
It's a general problem/pattern in eclipse code that the caller can modify arrays. the best solution would be to not use arrays at all. In this case if the caller modifies the error he would have a problem with ElementTree.ChildIDsCache.childPaths anyway. it's not the scope of this change to fix that

Reply:
I think the question is why a `public IPath[] rootPaths()` is needed at all...

Reply:
It seems the current state of the file do not contain any "rootPaths()" method so what has this to do with java 9? Even pre java 9 there is no one forcing you to return arrays ;-)